### PR TITLE
SALEOR-7900 - Grant refund: Send refund view

### PR DESCRIPTION
**Pull request by witoszekdev**
I want to merge this change because it adds send refund view to order details

### Screenshots

Requesting refund:

https://user-images.githubusercontent.com/4144459/199502124-233f4bb4-9302-42bf-93ed-3a01ce954540.mp4

Manual refunds:

https://user-images.githubusercontent.com/4144459/199502086-69748627-c7e6-4897-92d8-9ec62f659fac.mp4

### Pull Request Checklist

1. [ ] This code contains UI changes
2. [ ] All visible strings are translated with proper context including data-formatting
3. [ ] Attributes `[data-test-id]` are added for new elements
4. [ ] Changes are mentioned in the changelog
5. [ ] The changes are tested in different browsers and in light/dark mode

### Test environment config

API_URI=https://automation-dashboard.staging.saleor.cloud/graphql/
MARKETPLACE_URL=https://marketplace-gray.vercel.app/

### Do you want to run more stable tests?
Tests will be re-run only when the "run e2e" label is added.

1. [ ] stable
2. [ ] giftCard
3. [ ] category
4. [ ] collection
5. [ ] attribute
6. [ ] productType
7. [ ] shipping
8. [ ] customer
9. [ ] permissions
10. [ ] menuNavigation
11. [ ] pages
12. [ ] sales
13. [ ] vouchers
14. [ ] homePage
15. [ ] login
16. [x] orders
17. [ ] products
18. [ ] app
19. [ ] plugins
20. [ ] translations
21. [ ] navigation

CONTAINERS=1

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-7900 Send refund](https://app.clickup.com/t/2549495/SALEOR-7900)

File: src/orders/components/OrderSendRefundPage/messages.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  backlink: {
+    id: "rVIlBs",
+    defaultMessage: "Order #{orderNumber}",
+    description: "page header with order number",
+  },
+  backlinkNoNumber: {
+    id: "6u4K7e",
+    defaultMessage: "Order",
+    description: "page header",
+  },
```
Comment (orzechdev) [inline src/orders/components/OrderSendRefundPage/messages.ts:13]: I believe I saw it multiple times already

**Review by orzechdev — CHANGES_REQUESTED** [review]

File: src/orders/components/OrderSendRefundPage/OrderSendRefund.tsx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                <FormattedMessage
+                  {...refundPageMessages.balanceAfterRequests}
+                />
+              }
+            />
+            <CardContent>
+              <ul className={classes.dataList}>
+                {order.transactions.map(transaction => (
+                  <DataLine label={transaction.type}>
+                    {/* TODO: Replace with amounts stored in component */}
```
[the remark this answers is not in the document]
Comment (witoszekdev) [reply src/orders/components/OrderSendRefundPage/OrderSendRefund.tsx:129]: no, we don't need this actually

File: src/orders/components/OrderGrantRefundPage/components/RefundCard.tsx [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+          <span>
+            <FormattedMessage
+              {...grantRefundPageMessages.refundSelectedValue}
+            />
+          </span>
+          <span className={classes.totalMoney}>
+            {currency}&nbsp;
+            {getMoneyFormatted(locale, {
+              amount: totalSelectedPrice ?? 0,
+              currency: currency || "USD",
```
[the remark this answers is not in the document]
Comment (witoszekdev) [reply src/orders/components/OrderGrantRefundPage/components/RefundCard.tsx:91]: @orzechdev We just need something when the view is loading

File: src/orders/components/OrderSendRefundPage/messages.ts (hunk 1 above)
Comment (witoszekdev) [reply src/orders/components/OrderSendRefundPage/messages.ts:13]: ok, refactored it into one place

File: src/orders/components/OrderGrantRefundPage/OrderGrantRefundPage.tsx [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+              <ProductsCard
+                title={
+                  <FormattedMessage
+                    {...grantRefundPageMessages.unfulfilledProducts}
+                  />
+                }
+                lines={unfulfilledLines}
+              />
+              {order?.fulfillments?.map?.(fulfillment => (
+                <ProductsCard
```
[the remark this answers is not in the document]
Comment (witoszekdev) [reply src/orders/components/OrderGrantRefundPage/OrderGrantRefundPage.tsx:126]: When the page is loading, the lines and fulfillments array are empty, we don't want to show skeleton indefinitely, only when the page is actually loading - when there is nothing to grant refund for, we don't show anything and allow manual input

File: src/orders/components/OrderGrantRefundPage/components/RefundCard.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (orzechdev) [reply src/orders/components/OrderGrantRefundPage/components/RefundCard.tsx:91]: Can't we display a skeleton instead?

[the remark this answers is not in the document]
Comment (witoszekdev) [reply src/orders/components/OrderGrantRefundPage/components/RefundCard.tsx]: If data is loading skeleton is displayed (look a few lines below), I've removed the default USD currency

**Review by orzechdev — APPROVED** [review]

File: src/orders/components/OrderTransaction/components/TransactionEvents/components/PspReference.tsx [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  {
+    name: "PspReference",
+  },
+);
+
+enum RenderStep {
+  initial,
+  checkedMinimal,
+  final,
+}
```
Comment (Droniu) [inline src/orders/components/OrderTransaction/components/TransactionEvents/components/PspReference.tsx:24]: **nitpick**: I'd use union type here

File: src/orders/components/OrderTransaction/components/TransactionEvents/components/PspReference.tsx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  useEffect(() => {
+    if (ref.current && renderStep === RenderStep.final) {
+      // Forces re-calculation of width when text changes
+      setWidth(null);
+      setRenderStep(RenderStep.initial);
+    }
+  }, watch);
+
+  return { ref: onMount, width, renderStep };
+};
```
Comment (Droniu) [inline src/orders/components/OrderTransaction/components/TransactionEvents/components/PspReference.tsx:68]: **thought**: maybe if design requirements lead to such complex rendering cases we should reconsider design?

File: src/utils/errors/transaction.ts [hunk 7]
```diff
@@ -29,9 +31,42 @@ export function getOrderTransactionErrorMessage(
...
+  notFound: {
+    defaultMessage: "Cannot create transaction to non-existing order",
+    id: "aKSUWR",
+  },
+  incorrectCurrency: {
+    defaultMessage: "Invalid currency used to create transaction",
+    id: "/eWPsp",
+  },
+  unique: {
+    defaultMessage: "Transaction is not uniqe",
```
Comment (Droniu) [inline src/utils/errors/transaction.ts:53]: **typo:**
```suggestion
    defaultMessage: "Transaction is not unique",
```

**Review by Droniu — APPROVED** [review]

File: src/orders/components/OrderTransaction/components/TransactionEvents/components/PspReference.tsx (hunk 6 above)
Comment (witoszekdev) [reply src/orders/components/OrderTransaction/components/TransactionEvents/components/PspReference.tsx:68]: true, I'll leave it for now though

File: src/orders/components/OrderTransaction/components/TransactionEvents/components/PspReference.tsx (hunk 5 above)
Comment (witoszekdev) [reply src/orders/components/OrderTransaction/components/TransactionEvents/components/PspReference.tsx:24]: @Droniu why should I use a union type here? 🤔

File: src/orders/components/OrderSendRefundPage/components/ManualRefundCard.tsx [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  };
+
+  const handleChangeAmount: React.ChangeEventHandler<HTMLInputElement> = e => {
+    const value = parseFloat(e.target.value);
+    if (!Number.isNaN(value)) {
+      setAmount(value);
+    } else {
+      setAmount(undefined);
+    }
+  };
```
[the remark this answers is not in the document]
Comment (witoszekdev) [reply src/orders/components/OrderSendRefundPage/components/ManualRefundCard.tsx:79]: I think we should allow negative values here :D

File: src/orders/components/OrderTransaction/components/TransactionEvents/components/PspReference.tsx (hunk 5 above)
Comment (Droniu) [reply src/orders/components/OrderTransaction/components/TransactionEvents/components/PspReference.tsx:24]: It doesn't matter that much, it's a nitpick. Union types are generally easier to work with than enums (unique values, mixed types, subtyping) however both provide type safety which is the main reason for using them.

File: src/orders/components/OrderSendRefundPage/components/DataLineMoney.tsx [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+export const DataLineMoney: React.FC<DataLineMoneyProps> = ({ money }) => {
+  const { locale } = useLocale();
+  const classes = useDataLineMoneyStyles();
+
+  if (!money) {
+    return null;
+  }
+
+  const amount = getMoneyFormatted(locale, money);
```
[the remark this answers is not in the document]
Comment (witoszekdev) [reply src/orders/components/OrderSendRefundPage/components/DataLineMoney.tsx:19]: does this function also fill in the decimal place in currency? E.g `1 PLN` -> `1.00 PLN` There are currencies that don't have decimal places and some have more than 2

[the remark this answers is not in the document]
Comment (Droniu) [reply src/orders/components/OrderSendRefundPage/components/DataLineMoney.tsx:19]: Yes.

**Pull request merged**